default/AI/MilitaryAI.py

```diff
@@ -16,6 +17,34 @@
 totMilRating = 0
 num_milships = 0
 verbose_mil_reporting = False
+best_ship_rating_cache = {}  # indexed by turn, value is rating of that turn
+
+
+def cur_best_mil_ship_rating(include_designs=False):
+    """Find the best military ship we have available in this turn and return its rating.
```

Comment:
Keep in mind that the AI still has to build colony ships with an 8-10 turn leadtime and it would generally be best to not put off starting that until all the other ducks are 100% lined up in a row.  But that will be going away sometime and the calculations are tied into numerous other things with shorter timeframes, so I don't think we necessarily need to hold up this PR more on this point.  But do I want to note some things for continued discussion and future coding attention.   

For at least most of these uses the cur_best_mil_ship_rating is really being used as a rough measuring stick for on-call military power, and we should probably wind up shifting over to having the Military AI give some rating for how much firepower it has that is not required for defensive assignments.  

As for now, the use in AIFleetMission is not really one of "should I go this way" but "am I blocked by a threat so bad that I should just abort my mission".  The calc currently there is rather a rough hack at the decision and should surely get changed sometime.  In the meantime, basing the decision off of current ships is certainly the more conservative and probably the most effective. (Though did I mention that whole calc really needs attention?)  

The uses of cur_best_mil_ship_rating in ColonisationAI and in InvasionAI are for adjusting scoring of potential target planets, which have mixed timescale uses ranging from short-term decisions about which planet we should colonize/invade next, to medium-term decisions about "what are our prospects, how much investment should I put into building troops hips and outpost/colony ships"  As above, this should probably get shifted towards using a more direct assessment of on-call firepower, and we could give some thought to it taking into account warships on the build queue with only a couple turns left to complete.